# Data Exploratory Analysis of Flight Cancellations & Delays 

#### Author Name: Tejas Panhale 
#### Project Type: Data Analysis (EDA)  
#### Dataset: flights_sample_3m.csv

## Objective
Purpose The purpose of this project is to investigate the patterns of flight delays and cancellations to find operational inefficiencies, identify seasonal trends, and discover the primary reasons for delays.

This analysis aims to derive business insights that could help airlines improve punctuality and operational planning.

In [36]:
# Libraries required for the project are imported in this cell 

import pandas as pd

In [37]:
## Loading of the dataset saved locally

df = pd.read_csv("../data/raw/flights_sample_3m.csv")

# Display first few rows
df.head()


,FL_DATE,AIRLINE,AIRLINE_DOT,AIRLINE_CODE,DOT_CODE,FL_NUMBER,ORIGIN,ORIGIN_CITY,DEST,DEST_CITY,...,DIVERTED,CRS_ELAPSED_TIME,ELAPSED_TIME,AIR_TIME,DISTANCE,DELAY_DUE_CARRIER,DELAY_DUE_WEATHER,DELAY_DUE_NAS,DELAY_DUE_SECURITY,DELAY_DUE_LATE_AIRCRAFT
0,2019-01-09,United Air Lines Inc.,United Air Lines Inc.: UA,UA,19977,1562,FLL,"Fort Lauderdale, FL",EWR,"Newark, NJ",...,0.0,186.0,176.0,153.0,1065.0,NaN,NaN,NaN,NaN,NaN
1,2022-11-19,Delta Air Lines Inc.,Delta Air Lines Inc.: DL,DL,19790,1149,MSP,"Minneapolis, MN",SEA,"Seattle, WA",...,0.0,235.0,236.0,189.0,1399.0,NaN,NaN,NaN,NaN,NaN
2,2022-07-22,United Air Lines Inc.,United Air Lines Inc.: UA,UA,19977,459,DEN,"Denver, CO",MSP,"Minneapolis, MN",...,0.0,118.0,112.0,87.0,680.0,NaN,NaN,NaN,NaN,NaN
3,2023-03-06,Delta Air Lines Inc.,Delta Air Lines Inc.: DL,DL,19790,2295,MSP,"Minneapolis, MN",SFO,"San Francisco, CA",...,0.0,260.0,285.0,249.0,1589.0,0.0,0.0,24.0,0.0,0.0
4,2020-02-23,Spirit Air Lines,Spirit Air Lines: NK,NK,20416,407,MCO,"Orlando, FL",DFW,"Dallas/Fort Worth, TX",...,0.0,181.0,182.0,153.0,985.0,NaN,NaN,NaN,NaN,NaN


In [38]:
# Display last few rows

df.tail()

,FL_DATE,AIRLINE,AIRLINE_DOT,AIRLINE_CODE,DOT_CODE,FL_NUMBER,ORIGIN,ORIGIN_CITY,DEST,DEST_CITY,...,DIVERTED,CRS_ELAPSED_TIME,ELAPSED_TIME,AIR_TIME,DISTANCE,DELAY_DUE_CARRIER,DELAY_DUE_WEATHER,DELAY_DUE_NAS,DELAY_DUE_SECURITY,DELAY_DUE_LATE_AIRCRAFT
2999995,2022-11-13,American Airlines Inc.,American Airlines Inc.: AA,AA,19805,1522,JAX,"Jacksonville, FL",CLT,"Charlotte, NC",...,0.0,85.0,71.0,55.0,328.0,NaN,NaN,NaN,NaN,NaN
2999996,2022-11-02,American Airlines Inc.,American Airlines Inc.: AA,AA,19805,1535,ORD,"Chicago, IL",AUS,"Austin, TX",...,0.0,176.0,145.0,130.0,977.0,NaN,NaN,NaN,NaN,NaN
2999997,2022-09-11,Delta Air Lines Inc.,Delta Air Lines Inc.: DL,DL,19790,2745,HSV,"Huntsville, AL",ATL,"Atlanta, GA",...,0.0,55.0,50.0,28.0,151.0,0.0,36.0,0.0,0.0,0.0
2999998,2019-11-13,Republic Airline,Republic Airline: YX,YX,20452,6134,BOS,"Boston, MA",LGA,"New York, NY",...,0.0,88.0,77.0,50.0,184.0,NaN,NaN,NaN,NaN,NaN
2999999,2019-06-15,Southwest Airlines Co.,Southwest Airlines Co.: WN,WN,19393,2823,LGB,"Long Beach, CA",SJC,"San Jose, CA",...,0.0,70.0,63.0,52.0,324.0,NaN,NaN,NaN,NaN,NaN


## Dataset Overview

We will now examine:
- The number of rows and columns
- Data types
- Missing values
- Basic data summary


In [39]:
df.shape

(3000000, 32)

##### As per above output this dataset contains 3000000 rows and 32 columns. 

In [40]:
df.info() # To get the dataset column names and its datatypes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 32 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   FL_DATE                  object 
 1   AIRLINE                  object 
 2   AIRLINE_DOT              object 
 3   AIRLINE_CODE             object 
 4   DOT_CODE                 int64  
 5   FL_NUMBER                int64  
 6   ORIGIN                   object 
 7   ORIGIN_CITY              object 
 8   DEST                     object 
 9   DEST_CITY                object 
 10  CRS_DEP_TIME             int64  
 11  DEP_TIME                 float64
 12  DEP_DELAY                float64
 13  TAXI_OUT                 float64
 14  WHEELS_OFF               float64
 15  WHEELS_ON                float64
 16  TAXI_IN                  float64
 17  CRS_ARR_TIME             int64  
 18  ARR_TIME                 float64
 19  ARR_DELAY                float64
 20  CANCELLED                float64
 21  CANCELLA

#### To understand the completeness of our dataset, we calculate the total number of missing values in each column and sort them in descending order.

In [41]:
df.isnull().sum().sort_values(ascending=False) 

CANCELLATION_CODE          2920860
DELAY_DUE_LATE_AIRCRAFT    2466137
DELAY_DUE_CARRIER          2466137
DELAY_DUE_SECURITY         2466137
DELAY_DUE_NAS              2466137
DELAY_DUE_WEATHER          2466137
ARR_DELAY                    86198
ELAPSED_TIME                 86198
AIR_TIME                     86198
WHEELS_ON                    79944
TAXI_IN                      79944
ARR_TIME                     79942
WHEELS_OFF                   78806
TAXI_OUT                     78806
DEP_DELAY                    77644
DEP_TIME                     77615
CRS_ELAPSED_TIME                14
DEST_CITY                        0
CRS_DEP_TIME                     0
DEST                             0
DOT_CODE                         0
FL_NUMBER                        0
ORIGIN                           0
ORIGIN_CITY                      0
AIRLINE                          0
AIRLINE_DOT                      0
FL_DATE                          0
AIRLINE_CODE                     0
DIVERTED            

## Data Cleaning

Delay-related columns contain missing values because they are only populated when a delay occurs.

NOw we will replace NaN values in delay columns with 0 to correctly represent "no delay".


In [42]:
delay_columns = ['DELAY_DUE_LATE_AIRCRAFT',
                 'DELAY_DUE_CARRIER',
                 'DELAY_DUE_SECURITY',
                 'DELAY_DUE_NAS',
                 'DELAY_DUE_WEATHER']
    
df[delay_columns] = df[delay_columns].fillna(0)

df[delay_columns].isnull().sum()

DELAY_DUE_LATE_AIRCRAFT    0
DELAY_DUE_CARRIER          0
DELAY_DUE_SECURITY         0
DELAY_DUE_NAS              0
DELAY_DUE_WEATHER          0
dtype: int64

## Cancellation Analysis

Now we will calculate the total number and percentage of cancelled flights.


In [43]:
total_flights = len(df)

cancelled_flights = df['CANCELLED'].sum()

cancellation_rate = (cancelled_flights / total_flights) * 100

print(f"""
Flight Cancellation Summary
------------------------------
Total Flights      : {total_flights:,}
Cancelled Flights  : {cancelled_flights:,}
Cancellation Rate  : {cancellation_rate:.2f}%
""")



Flight Cancellation Summary
------------------------------
Total Flights      : 3,000,000
Cancelled Flights  : 79,140.0
Cancellation Rate  : 2.64%



### Insight

The overall cancellation rate is 2.64%, meaning that approximately 97% of flights operate successfully.

This suggests that cancellations are relatively uncommon compared to delays, and operational inefficiency will show as a delay rather than as a cancellation.
